curriculum/challenges/english/25-front-end-development/lab-nth-fibonacci-number-generator/66d9af3897e7d75a895b72c2.md

```diff
@@ -0,0 +1,116 @@
+---
+id: 66d9af3897e7d75a895b72c2
+title: Build the nth Fibonacci number generator
+challengeType: 14
+dashedName: lab-nth-fibonacci-number-generator
+---
+
+# --description--
+
+In this lab you will use a dynamic programming approach to compute and return the `n`-th number from the Fibonacci sequence, in which each number is the sum of the two preceding numbers.
+
+**Objective:** Fulfill the user stories below and get all the tests to pass to complete the lab. 
+
+**User Stories:**
+
+1. You should define an array named `sequence` in the global scope. It should contain the number `0` and `1`.
+1. You should create a function named `fibonacci`.
+1. The `fibonacci` function should take a single parameter, which is a positive integer.
+1. If the input is not a positive integer, the function should return `"Invalid input"`.
+1. A call to `fibonacci(n)` should use a dynamic programming approach to compute and return the `n`-th number from the Fibonacci sequence.
+1. Each computed number at the position `n` in the Fibonacci sequence should be stored at the index `n - 1` of the `sequence` array.
+1. If the input number requests an already computed number, no new computations should happen and the number should be retrieved from the `sequence` array.
+
+# --hints--
+
+You should have a `sequence` array.
+
+```js
+assert.isArray(sequence);
+```
+
+The `sequence` array should contain `0` and `1` at definition.
+
+```js
+assert.match(__helpers.removeJSComments(code), /(let|const)\s+sequence\s*=\s*\[\s*0\s*,\s*1\s*\]\s*;?/)
+```
+
+You should create a `fibonacci` function.
+
+```js
+assert.isFunction(fibonacci);
+```
+
+If the `fibonacci` function receives an argument that is not a positive integer it should return `"Invalid input"`.
+
+```js
+assert.strictEqual(fibonacci(3.5), "Invalid input")
+assert.strictEqual(fibonacci(0), "Invalid input")
+assert.strictEqual(fibonacci(-1), "Invalid input")
+assert.strictEqual(fibonacci("three"), "Invalid input")
+```
+
+`fibonacci(1)` should return `0`.
+
+```js
+assert.strictEqual(fibonacci(1), 0)
+
+```
+
+`fibonacci(32)` should return `1346269`.
+
+```js
+assert.strictEqual(fibonacci(32), 1346269)
+
+```
+
+`fibonacci(5)` should return `3`.
+
+```js
+assert.strictEqual(fibonacci(5), 3);
+
+```
+
+`sequence` should contain numbers up to the highest call so far.
+
+```js
+sequence.length = 2
+fibonacci(2)
+assert.lengthOf(sequence, 2)
+fibonacci(5)
+assert.lengthOf(sequence, 5)
+fibonacci(3)
+assert.lengthOf(sequence, 5)
```

Comment:
How can two different values have the same output? This doesn't make sense.

Reply:
I see, that's still a little confusing. Could we add an example for this?

Reply:
Where would you add the example? Before user stories?

Reply:
if the camper compute the Fibonacci sequence as requested, outside the function, it doesn't need an example. This is checking that if the number already exist in the sequence, then no new numbers are added to the sequence

Reply:
Okay, I will resolve this one.